### ISSUE TITLE
[feature request] Adapt to a private image repository with prefixes such as harbor

### ISSUE THREAD

**Issue by cxwen** [dragonflyoss/Dragonfly#299]
---
name: Feature Request
about: Suggest your excellent idea for this project
---

## Why you need it?

如果使用harbor等私有镜像仓库，一般都带有镜像仓库域名或IP的前缀，例如：
```
docker pull hub.aaa.com/xxx/yyyy:v1.0
```
但是目前dragonfly不支持这种带有前缀的镜像拉取方式，正如下面项目文档中所说：

> 直接使用`docker pull imageName`下载镜像即可。

> **注意**：镜像名称不要包含镜像仓库地址，因为仓库域名已经由`df-daemon`的启动参数`--registry`指定。

但这样对于已经使用harbor等镜像仓库的项目来说，引入dragonfly还需要把所有使用到的镜像名称中的前缀去掉，会对项目使用dragonfly带来很大麻烦。

## How it could be?
在dfdaemon改URL之前做一个判断，如果镜像名带有前缀，直接使用它的前缀；没有才自动为其添加URL.

**Comment by allencloud** [comment]
I think this is a practical feature request indeed. WDYT? @Starnop @zhouhaibing089 @lowzj @garfield009

**Comment by Starnop** [comment]
Yeah, I will fix it.

**Comment by haosdent** [comment]
@Starnop @allencloud docker daemon would request hub.aaa.com always when `docker pull hub.aaa.com/xxx/yyyy:v1.0` instead of sending requests to `registry mirrors`, right? So this issue didn't resolve the problem?

Refer to the code at https://github.com/docker/docker-ce/blob/master/components/engine/registry/service_v2.go#L12-L17

```
	if hostname == DefaultNamespace || hostname == IndexHostname {
		// v2 mirrors
		for _, mirror := range s.config.Mirrors {
			if !strings.HasPrefix(mirror, "http://") && !strings.HasPrefix(mirror, "https://") {
				mirror = "https://" + mirror
			}
```

**Comment by allencloud** [comment]
Totally agree with you. @haosdent 
I am afraid that we cannot handle this case with the normal way.

**Comment by haosdent** [comment]
I have a simple change at https://github.com/moby/moby/pull/38500 , but not sure if it satisfies upstream.

**Comment by lowzj** [comment]
This's resolved by #389
